corehq/apps/smsforms/models.py

```diff
@@ -237,3 +240,82 @@ def move_to_next_action(self):
         while self.current_action_is_a_reminder and self.current_action_due < utcnow():
             self.current_reminder_num += 1
             self.set_current_action_due_timestamp()
+
+
+class XFormsSessionSynchronization:
+
+    @classmethod
+    def claim_phone_number_for_session(cls, phone_number, session):
+        """
+        This is a non-blocking acquire.
+
+        The session that claims a phone number must also release it when it's over.
+        Returns True if it was able to claim the number.
+        """
+        with cls._critical_section(phone_number):
+            if cls.could_maybe_claim_phone_number_for_session_id(phone_number, session.session_id):
+                cls._set_running_session_info_for_phone_number(
+                    phone_number,
+                    RunningSessionInfo(session.session_id, session.connection_id),
+                    # We aren't relying on the expiry here to free up the session: we manually release it.
+                    # Still, there is no situation where we'd want to keep this longer than that.
+                    session.expire_after * 60,
+                )
+                return True
+            else:
+                return False
+
+    @classmethod
+    def could_maybe_claim_phone_number_for_session_id(cls, phone_number, session_id):
+        """
+        Check if there's another session running for heuristic purposes
```

Comment:
I'm a bit confused by the purpose of this function, what do you mean by heuristic purposes?

Reply:
I just mean that you can use it to see if you _would be able to_ claim it right now, but that there's no guarantee that if you try to claim it, even in the next line, that this function would have predicted correctly, because it's not part of any locking that makes it atomic. The actual claim method on the other hand is atomic because of the critical section block.

So for example, in the `handle_survey_actions` command, I call this function to see if it's even worth creating a task for it. So it's a "heuristic" in the sense that it's a pretty good signal you can use to predict whether it's worth _trying_ to claim the channel for a session, but that it's only going to be accurate most of the time so it's not good for making critical decisions on. So "this said I _could_ claim the channel, so I'm going to assume I _will be able to_ claim the channel", that would be a bad assumption; whereas "this said I _could_ claim it, so I'm going to spin up a little work to try and claim it" is a good way to use it.

Reply:
Aha, thanks, that makes sense. What would you think of calling this something like `phone_number_eligible_for_session_id`?

Reply:
The code changed since your first comment, but I took the spirit of it and renamed `could_maybe_claim_channel_for_session` to `channel_is_available_for_session`

Reply:
1f37125